ragmac_xdem/temporal.py

```diff
@@ -310,3 +319,194 @@ def ma_linreg(
         slope *= 365.25
 
     return slope, intercept, detrended_std
+
+
+"""
+@author: friedrichknuth
+"""
+
+def remove_nan_from_training_data(X_train, y_train_masked_array):
+    array = y_train_masked_array.data
+    mask = ~np.ma.getmaskarray(y_train_masked_array)                    
+    X_train = X_train[mask]
+    y_train = y_train_masked_array[mask]
+    return X_train, y_train
+
+def mask_low_count_pixels(ma_stack, n_thresh = 3):
+    count = np.ma.masked_equal(ma_stack.count(axis=0), 0).astype(np.uint16).data
+    valid_mask_2D = (count >= n_thresh)
+    valid_data = ma_stack[:, valid_mask_2D]
+    return valid_data, valid_mask_2D
+
+def create_prediction_timeseries(start_date = '2000-01-01',
+                                 end_date = '2023-01-01',
+                                 dt ='M'):
+    #M  = monthly frequency
+    #3M = every 3 months
+    #6M = every 6 months
+    d = pd.date_range(start_date,end_date,freq=dt)
+    X = d.to_series().apply([utils.date_time_to_decyear]).values.squeeze()
+    return X
+
+def linreg_fit(X_train,
+               y_train,
+               method='TheilSen'):
+    
+    if method=='Linear':
+        m = linear_model.LinearRegression()
+        m.fit(X_train.squeeze()[:,np.newaxis], y_train.squeeze())
+        slope = m.coef_
+        intercept = m.intercept_
+#         prediction = m.predict(X.squeeze()[:,np.newaxis])
+
+    if method=='TheilSen':
+        m = linear_model.TheilSenRegressor()
+        m.fit(X_train.squeeze()[:,np.newaxis], y_train.squeeze())
+        slope = m.coef_
+        intercept = m.intercept_
+
+    if method=='RANSAC':
+        m = linear_model.RANSACRegressor()
+        m.fit(X_train.squeeze()[:,np.newaxis], y_train.squeeze())
+        slope = m.estimator_.coef_
+        intercept = m.estimator_.intercept_
+
+    return slope[0], intercept
+
+def linreg_run(args):
+    X_train, y_train_masked_array, method = args
+    
+    X_train, y_train = remove_nan_from_training_data(X_train, y_train_masked_array)
+    slope, intercept = linreg_fit(X_train,
+                                  y_train,
+                                  method='Linear')
+    
+    return slope, intercept
+
+def linreg_predict(args):
+    slope, x, intercept = args
+    prediction = slope * x + intercept
+    return prediction
+
+def linreg_predict_parallel(slope,X,intercept):
+    pool = mp.Pool(processes=psutil.cpu_count(logical=True))
```

Comment:
would be good to make it an optional argument, and pass the option in the main scripts.